src/contracts/entrypoint.rs

```diff
@@ -0,0 +1,238 @@
+use std::str::FromStr;
+use std::sync::Arc;
+
+use anyhow;
+use ethers::abi::AbiDecode;
+use ethers::prelude::ContractError;
+use ethers::providers::Middleware;
+use ethers::types::{Address, Bytes};
+use regex::Regex;
+use serde::Deserialize;
+
+use super::gen::entry_point_api::{self, UserOperation};
+use super::gen::EntryPointAPI;
+
+pub struct EntryPoint<M: Middleware> {
+    provider: Arc<M>,
+    entry_point_address: Address,
+    api: EntryPointAPI<M>,
+}
+
+impl<M: Middleware + 'static> EntryPoint<M> {
+    pub fn new(provider: Arc<M>, entry_point_address: Address) -> Self {
+        let api = EntryPointAPI::new(entry_point_address, provider.clone());
+        Self {
+            provider,
+            entry_point_address: entry_point_address,
+            api,
+        }
+    }
+
+    pub fn provider(&self) -> Arc<M> {
+        self.provider.clone()
+    }
+
+    pub fn entry_point_address(&self) -> Address {
+        self.entry_point_address
+    }
+
+    fn deserialize_error_msg(
+        err_msg: &str,
+    ) -> Result<entry_point_api::EntryPointAPIErrors, EntryPointErr> {
+        JsonRpcError::from_str(&err_msg)
+            .map_err(|_| {
+                EntryPointErr::DecodeErr(format!(
+                    "{:?} is not a valid JsonRpcError message",
+                    err_msg
+                ))
+            })
+            .and_then(|json_error| {
+                json_error.data.ok_or_else(|| {
+                    EntryPointErr::DecodeErr("{:?} doesn't have valid data field".to_string())
+                })
+            })
+            .and_then(|data: String| {
+                AbiDecode::decode_hex(data).map_err(|_| {
+                    EntryPointErr::DecodeErr(format!(
+                        "{:?} data field could not be deserialize to EntryPointAPIErrors",
+                        err_msg
+                    ))
+                })
+            })
+    }
+
+    pub async fn get_user_op_hash<U: Into<UserOperation>>(
```

Comment:
I do not think this function is needed; it's implemented in user_operation.rs (and no need to make another RPC call to provider/client).